### ISSUE TITLE
Any preprocess needed by problem “wmt_ende_tokens_32k”?

### ISSUE THREAD

**Issue by ZhbChopin** [tensorflow/tensor2tensor#84]
I trained on 4 NVIDIA Tesla  M40 with the “transformer_base“ parameters:
And I get loss as below:

![image](https://user-images.githubusercontent.com/28777721/27781578-0d5693fa-6001-11e7-891e-0ecc293c805a.png)

My test set is [wmt14](http://www.statmt.org/wmt14/test-filtered.tgz).
 the eval BLEU is about 21.
Is it normal?  
And any preprocess needed?

**Comment by mehmedes** [comment]
As a preprocessing step, the input to be translated needs to be  tokenized:
```
$moses-scripts/scripts/tokenizer/tokenizer.perl -l $SRC
```
and after tokenization the bpe file needs to be applied to the tokenized input:
```
$subword-nmt/apply_bpe.py -c bpe.32000 
```
After translation, t2t will have created a tab-delimited file containing source text and translation.
Make sure sure to extract the column with the translation:
```
cut -d'	' -f1
```
Then run:
```
sed 's/\@\@ //g' 
```
As translating with the bpe model produces tokenized output, there's no need to run the tokenizer in the end.
Before you run
```
./moses-scripts/scripts/generic/multi-bleu.perl
```
to evaluate the blue score, make sure to apply: 
```
perl -ple 's{(\S)-(\S)}{$1 ##AT##-##AT## $2}g'
```
to the translation and the reference text.

**Comment by ZhbChopin** [comment]
@mehmedes thanks!   It's greatly helpful.
I made a mistake in my title. Indeed the problem is "wmt_ende_tokens_32k"!  
I have changed it

**Comment by mehmedes** [comment]
In that case you can use the input without preprocessing.
Make sure to [average](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/avg_checkpoints.py) your checkpoints and run  [get_ende.bleu.sh](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/get_ende_bleu.sh) to postprocess the output and evaluate bleu.

**Comment by ZhbChopin** [comment]
@mehmedes thanks, I got it. 
I didn't run the tokenizer before run multi-bleu.perl
And I used only one checkpoint to evaluate bleu

**Comment by colmantse** [comment]
Dear all,

I am interested in the preprocessing/padding scripts in tensor2tensor, any idea where I could locate them? I checked the tmp folders that contents are downloaded into with t2t-datagen, it seems they are sentences of variable lengths. They are at their initial states so my best bet is there has to be preprocessing somewhere during initial training.

Best,
Colman

**Comment by lukaszkaiser** [comment]
Answered on #221 , hope that helps :).

**Comment by colmantse** [comment]
It helps definitely! Thanks
